### PR TITLE
Upgrade PHPStan to level 3

### PR DESCRIPTION
**Pull request by lruozzi9**
| Q               | A                                                            |
|-----------------|--------------------------------------------------------------|
| Branch?         | 1.12 |
| Bug fix?        | no                                                      |
| New feature?    | no                                                      |
| BC breaks?      | no                                                      |
| Deprecations?   | no |
| Related tickets | partially #14019                      |
| License         | MIT                                                          |

Written during the journey to the SyliusCon, so lit 🔥 😆

### CODE REVIEW

File: src/Sylius/Bundle/CoreBundle/Form/Extension/CountryTypeExtension.php [hunk 1]
```diff
@@ -80,9 +80,6 @@ private function getCountryName(string $code): string
         return Countries::getName($code);
     }
 
-    /**
-     * @return array|CountryInterface[]
-     */
```
Comment (jakubtobiasz) [inline src/Sylius/Bundle/CoreBundle/Form/Extension/CountryTypeExtension.php:85]: Why did you remove `CountryInterface[]` type-hint?

File: src/Sylius/Bundle/CoreBundle/Form/Extension/LocaleTypeExtension.php [hunk 2]
```diff
@@ -66,9 +66,6 @@ private function getLocaleName(string $code): ?string
         return Locales::getName($code);
     }
 
-    /**
-     * @return array|LocaleInterface[]
-     */
```
Comment (jakubtobiasz) [inline src/Sylius/Bundle/CoreBundle/Form/Extension/LocaleTypeExtension.php:71]: Why did you remove `LocaleInterface[]` type-hint?

File: src/Sylius/Bundle/PayumBundle/Request/GetStatus.php [hunk 3]
```diff
@@ -20,6 +20,7 @@ class GetStatus extends BaseGetStatus
 {
     /**
      * @psalm-suppress NonInvariantDocblockPropertyType
+     * @phpstan-ignore-next-line
```
Comment (jakubtobiasz) [inline src/Sylius/Bundle/PayumBundle/Request/GetStatus.php:23]: What kind of error did you get that you need to ignore it?

File: src/Sylius/Bundle/ShopBundle/Router/LocaleStrippingRouter.php [hunk 4]
```diff
@@ -57,11 +57,13 @@ public function getRouteCollection(): RouteCollection
         return $this->router->getRouteCollection();
     }
 
-    public function warmUp($cacheDir): void
+    public function warmUp($cacheDir): array
```
Comment (jakubtobiasz) [inline src/Sylius/Bundle/ShopBundle/Router/LocaleStrippingRouter.php:61]: As it returns an array of strings, add `string[]` as a return type in the phpdoc, please.

File: src/Sylius/Component/Core/Model/CatalogPromotion.php [hunk 5]
```diff
@@ -41,6 +42,7 @@ public function __construct()
      */
     public function getChannels(): Collection
     {
+        /** @phpstan-ignore-next-line */
         return $this->channels;
```
Comment (jakubtobiasz) [inline src/Sylius/Component/Core/Model/CatalogPromotion.php:46]: Why did you need to ignore this?

File: src/Sylius/Component/Core/Model/Order.php [hunk 6]
```diff
@@ -208,6 +209,7 @@ public function getItemUnitsByVariant(ProductVariantInterface $variant): Collect
      */
     public function getPayments(): Collection
     {
+        /** @phpstan-ignore-next-line */
         return $this->payments;
```
Comment (jakubtobiasz) [inline src/Sylius/Component/Core/Model/Order.php:213]: As above, why we need to ignore this?

File: src/Sylius/Component/Core/Model/OrderInterface.php [hunk 7]
```diff
@@ -132,6 +132,8 @@ public function setCreatedByGuest(bool $createdByGuest): void;
      * @psalm-return Collection<array-key, OrderItemInterface>
      *
      * @psalm-suppress ImplementedReturnTypeMismatch
+     *
+     * @phpstan-ignore-next-line
```
Comment (jakubtobiasz) [inline src/Sylius/Component/Core/Model/OrderInterface.php:136]: As above, why ignore?

File: src/Sylius/Component/Core/Model/Product.php [hunk 8]
```diff
@@ -149,6 +149,7 @@ public function hasTaxon(TaxonInterface $taxon): bool
      */
     public function getChannels(): Collection
     {
+        /** @phpstan-ignore-next-line */
         return $this->channels;
```
Comment (jakubtobiasz) [inline src/Sylius/Component/Core/Model/Product.php:153]: Why ignore?

File: src/Sylius/Component/Core/Model/ProductVariant.php [hunk 9]
```diff
@@ -291,6 +292,7 @@ public function getAppliedPromotionsForChannel(ChannelInterface $channel): Colle
      */
     public function getImages(): Collection
     {
+        /** @phpstan-ignore-next-line */
         return $this->images;
```
Comment (jakubtobiasz) [inline src/Sylius/Component/Core/Model/ProductVariant.php:296]: Why ignore?

File: src/Sylius/Component/Core/Model/Promotion.php [hunk 10]
```diff
@@ -41,6 +41,7 @@ public function __construct()
      */
     public function getChannels(): Collection
     {
+        /** @phpstan-ignore-next-line */
         return $this->channels;
```
Comment (jakubtobiasz) [inline src/Sylius/Component/Core/Model/Promotion.php:45]: Why ignore?

File: src/Sylius/Component/Core/Model/ShippingMethod.php [hunk 11]
```diff
@@ -70,6 +70,7 @@ public function setTaxCategory(?TaxCategoryInterface $category): void
      */
     public function getChannels(): Collection
     {
+        /** @phpstan-ignore-next-line */
         return $this->channels;
```
Comment (jakubtobiasz) [inline src/Sylius/Component/Core/Model/ShippingMethod.php:74]: Why ignore?

File: src/Sylius/Component/Core/Model/Shipment.php [hunk 12]
```diff
@@ -44,6 +45,7 @@ public function __construct()
 
     public function getOrder(): ?BaseOrderInterface
     {
+        Assert::nullOrIsInstanceOf($this->order, OrderInterface::class);
```
Comment (diimpp) [inline src/Sylius/Component/Core/Model/Shipment.php:48]: I was wondering for long time, but is it no possible to cover all this cases with extended resource bundle models with phpstan plugin somehow?

File: src/Sylius/Component/Shipping/Resolver/DefaultShippingMethodResolver.php [hunk 13]
```diff
@@ -30,7 +31,9 @@ public function getDefaultShippingMethod(ShipmentInterface $shipment): ShippingM
         if (empty($shippingMethods)) {
             throw new UnresolvedDefaultShippingMethodException();
         }
+        $shippingMethod = $shippingMethods[0];
+        Assert::isInstanceOf($shippingMethod, ShippingMethodInterface::class);
 
-        return $shippingMethods[0];
+        return $shippingMethod;
```
Comment (diimpp) [inline src/Sylius/Component/Shipping/Resolver/DefaultShippingMethodResolver.php:37]: Missing empty line before `return`

File: src/Sylius/Bundle/CoreBundle/Form/Extension/CountryTypeExtension.php (hunk 1 above)
Comment (lruozzi9) [reply src/Sylius/Bundle/CoreBundle/Form/Extension/CountryTypeExtension.php:85]: It was not right, the method returns a string[]. It could be automatically recognized by PHPStan but I anyway added it to the doc.

File: src/Sylius/Bundle/CoreBundle/Form/Extension/LocaleTypeExtension.php (hunk 2 above)
Comment (lruozzi9) [reply src/Sylius/Bundle/CoreBundle/Form/Extension/LocaleTypeExtension.php:71]: It was not right, the method returns a string[]. It could be automatically recognized by PHPStan but I anyway added it to the doc.

File: src/Sylius/Component/Core/Model/Shipment.php (hunk 12 above)
Comment (diimpp) [reply src/Sylius/Component/Core/Model/Shipment.php:48]: To clarify the point: There is valid BaseOrderInterface return type hint, but phpstan still forces to check against current OrderInterface.
Might there be an option to make phpstan aware of resource bundle inheritance scheme somehow and skip this assert.

But this is purely off topic.

File: src/Sylius/Bundle/PayumBundle/Request/GetStatus.php (hunk 3 above)
Comment (lruozzi9) [reply src/Sylius/Bundle/PayumBundle/Request/GetStatus.php:23]:
```
PHPDoc type string of property Sylius\Bundle\PayumBundle\Request\GetStatus::$status is not covariant with PHPDoc type int of overridden property
         Payum\Core\Request\BaseGetStatus::$status.
         💡 You can fix 3rd party PHPDoc types with stub files:
            https://phpstan.org/user-guide/stub-files
```
Payum annotates the property as an int, but it is used as a string here. Obviously, it is not so easy to change here to an int and it is not the purpose of this PR.

File: src/Sylius/Component/Core/Model/CatalogPromotion.php (hunk 5 above)
Comment (lruozzi9) [reply src/Sylius/Component/Core/Model/CatalogPromotion.php:46]:
```
Method Sylius\Component\Core\Model\CatalogPromotion::getChannels() should return Doctrine\Common\Collections\Collection<(int|string),
         Sylius\Component\Channel\Model\ChannelInterface> but returns Doctrine\Common\Collections\Collection<(int|string), Sylius\Component\Core\Model\ChannelInterface>.
```
Obviously here it is right to return a list of core channels, it sounds like a PHPStan bug...

File: src/Sylius/Component/Core/Model/Order.php (hunk 6 above)
Comment (lruozzi9) [reply src/Sylius/Component/Core/Model/Order.php:213]:
```
Method Sylius\Component\Core\Model\Order::getPayments() should return Doctrine\Common\Collections\Collection<(int|string),
         Sylius\Component\Payment\Model\PaymentInterface> but returns Doctrine\Common\Collections\Collection<(int|string), Sylius\Component\Core\Model\PaymentInterface>.
```
Obviously here it is right to return a list of core payment, it sounds like a PHPStan bug...

File: src/Sylius/Component/Core/Model/OrderInterface.php (hunk 7 above)
Comment (lruozzi9) [reply src/Sylius/Component/Core/Model/OrderInterface.php:136]:
```
Return type (Doctrine\Common\Collections\Collection<(int|string), Sylius\Component\Core\Model\OrderItemInterface>) of method
         Sylius\Component\Core\Model\OrderInterface::getItems() should be compatible with return type (Doctrine\Common\Collections\Collection<(int|string),
         Sylius\Component\Order\Model\OrderItemInterface>) of method Sylius\Component\Order\Model\OrderInterface::getItems()
```
Obviously here it is right to return a list of core order items, it sounds like a PHPStan bug...

File: src/Sylius/Component/Core/Model/Product.php (hunk 8 above)
Comment (lruozzi9) [reply src/Sylius/Component/Core/Model/Product.php:153]:
```
Method Sylius\Component\Core\Model\Product::getChannels() should return Doctrine\Common\Collections\Collection<(int|string),
         Sylius\Component\Channel\Model\ChannelInterface> but returns Doctrine\Common\Collections\Collection<(int|string), Sylius\Component\Core\Model\ChannelInterface>.
```
Obviously here it is right to return a list of core channels, it sounds like a PHPStan bug...

File: src/Sylius/Component/Core/Model/ProductVariant.php (hunk 9 above)
Comment (lruozzi9) [reply src/Sylius/Component/Core/Model/ProductVariant.php:296]:
```
Method Sylius\Component\Core\Model\ProductVariant::getImages() should return Doctrine\Common\Collections\Collection<(int|string),
         Sylius\Component\Core\Model\ImageInterface> but returns Doctrine\Common\Collections\Collection<(int|string), Sylius\Component\Core\Model\ProductImageInterface>.
```
Obviously here it is right to return a list of core product images, it sounds like a PHPStan bug...

File: src/Sylius/Component/Core/Model/Promotion.php (hunk 10 above)
Comment (lruozzi9) [reply src/Sylius/Component/Core/Model/Promotion.php:45]:
```
Method Sylius\Component\Core\Model\Promotion::getChannels() should return Doctrine\Common\Collections\Collection<(int|string),
         Sylius\Component\Channel\Model\ChannelInterface> but returns Doctrine\Common\Collections\Collection<(int|string), Sylius\Component\Core\Model\ChannelInterface>.
```
Obviously here it is right to return a list of core channels, it sounds like a PHPStan bug...

File: src/Sylius/Component/Core/Model/ShippingMethod.php (hunk 11 above)
Comment (lruozzi9) [reply src/Sylius/Component/Core/Model/ShippingMethod.php:74]:
```
Method Sylius\Component\Core\Model\ShippingMethod::getChannels() should return Doctrine\Common\Collections\Collection<(int|string),
         Sylius\Component\Channel\Model\ChannelInterface> but returns Doctrine\Common\Collections\Collection<(int|string), Sylius\Component\Core\Model\ChannelInterface>.
```
Obviously here it is right to return a list of core channels, it sounds like a PHPStan bug...

File: src/Sylius/Component/Core/Model/Shipment.php (hunk 12 above)
Comment (lruozzi9) [reply src/Sylius/Component/Core/Model/Shipment.php:48]: I think this could also be due to some wrong property type. For example, in this class, the order property is annotated as a `BaseOrderInterface|null`, but this is not right, it should be `OrderInterface|null`. Anyway, moving on to PHP 8.0 and 8.1 makes big huge forward in typization these annotations can be changed in a more strict way, but obviously, it is not the purpose of this PR.

**Comment by lruozzi9** [comment]
Sorry for the delay 🤓

**Comment by jakubtobiasz** [comment]
@lruozzi9 no reason to apologize for delay 😄. We appreciate the work you do. Tomorrow I will come back to review the PR again to accelerate the whole process of reviewing your job 😎.

**Comment by lruozzi9** [comment]
Hello, any update?

File: src/Sylius/Bundle/CoreBundle/Form/Extension/CountryTypeExtension.php (hunk 1 above)
Comment (jakubtobiasz) [reply src/Sylius/Bundle/CoreBundle/Form/Extension/CountryTypeExtension.php:85]: Great, it is used by IDE too, so when you do foreach on an array you get type-hint :D.

**Review by jakubtobiasz — APPROVED** [review]

**Comment by jakubtobiasz** [comment]
Hi @lruozzi9!
Sorry for the delay, busy schedule 😅. I approved these changes, I'll ping the team to gather more feedback. I bet (at the least) in the next week this PR should be merged :).

**Comment by lruozzi9** [comment]
Ok @jakubtobiasz no problem 😉

File: src/Sylius/Component/Core/Model/Address.php [hunk 14]
```diff
@@ -28,6 +29,7 @@ public function getCustomer(): ?BaseCustomerInterface
 
     public function setCustomer(?BaseCustomerInterface $customer): void
     {
+        Assert::nullOrIsInstanceOf($customer, CustomerInterface::class);
```
Comment (coldic3) [inline src/Sylius/Component/Core/Model/Address.php:32]: Shouldn't it be changed in phpdoc instead, above the `Address::customer` property?

File: src/Sylius/Bundle/AdminBundle/Controller/DashboardController.php [hunk 15]
```diff
@@ -71,9 +72,15 @@ public function getRawData(Request $request): Response
     private function findChannelByCodeOrFindFirst(?string $channelCode): ?ChannelInterface
     {
         if (null !== $channelCode) {
-            return $this->channelRepository->findOneByCode($channelCode);
+            $channel = $this->channelRepository->findOneByCode($channelCode);
+            Assert::nullOrIsInstanceOf($channel, ChannelInterface::class);
```
Comment (coldic3) [inline src/Sylius/Bundle/AdminBundle/Controller/DashboardController.php:76]: I'm just wondering if there are any pros to having assertions in such cases. We are fairly sure it is a `ChannelInterface` because we use a channel repository. So maybe the `@var ChannelInterface|null` annotation is enough. Thanks to this we get rid of some extra if statements. WDYT?

**Review by NoResponseMate — APPROVED** [review]

File: src/Sylius/Bundle/AdminBundle/Controller/DashboardController.php (hunk 15 above)
Comment (diimpp) [reply src/Sylius/Bundle/AdminBundle/Controller/DashboardController.php:76]: > https://github.com/phpstan/phpstan-doctrine
> Interprets `EntityRepository<MyEntity>` correctly in phpDocs for further type inference of methods called on the repository.
> Recognizes magic findBy*, findOneBy* and countBy* methods on EntityRepository.

Do you guys think it does what I think it does? That would remove need to define magic methods for typehint, like so
https://github.com/Sylius/Sylius/blob/2818ace6223948f9b91a95aa3a41c0b6bea0d718/src/Sylius/Bundle/ChannelBundle/Doctrine/ORM/ChannelRepository.php#L32-L35

and if we're to do a `phpstan-sylius-resource` extension, which will read current channel class from resource configuration (They do read doctrine mappings as well https://github.com/phpstan/phpstan-doctrine/tree/1.3.x/src/Doctrine/Mapping) and supply actual class to this `EntityRepository<MyEntity>` value, then we shouldn't need to add asserts or @var each time repository is used.

In theory. :)

**Review by TheMilek — APPROVED** [review]

**Comment by coldic3** [comment]
Thank you, Lorenzo! :1st_place_medal:

File: src/Sylius/Bundle/AdminBundle/Controller/DashboardController.php (hunk 15 above)
Comment (lruozzi9) [reply src/Sylius/Bundle/AdminBundle/Controller/DashboardController.php:76]: Sorry, I didn't answer this question... Of course, there is some misconfiguration in the type documented in the methods or in the way PHPStan, Psalm, and/or PHPStorm recognize these objects, but it is not the purpose of this PR. Probably another issue should be opened and this discussion should be moved there.
